### DISCUSSION TITLE
Import error

### DISCUSSION THREAD

**Discussion by GengXia0406** [QMCSoftware/QMCSoftware#244] [Q&A]
Hi there

I successfully installed QMC but was having trouble importing it into Notebook. Below shows the error that I received. I would appreciate if anyone can provide any suggestions.

Thanks in advance
Best regards

Geng Xia
Postdoctoral Associate| National Wind Technology Center
National Renewable Energy Laboratory (NREL)

---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
<ipython-input-2-db61659dc1e0> in <module>
     22 import glob
     23 import pathlib
---> 24 from qmcpy.integrand import *

~/.conda-envs/playground/lib/python3.6/site-packages/qmcpy/__init__.py in <module>
----> 1 from .discrete_distribution import *
      2 from .true_measure import *
      3 from .integrand import *
      4 from .stopping_criterion import *
      5 

~/.conda-envs/playground/lib/python3.6/site-packages/qmcpy/discrete_distribution/__init__.py in <module>
      2 from .iid_std_gaussian import IIDStdGaussian
      3 from .lattice import Lattice
----> 4 from .sobol import Sobol, DigitalNet
      5 from .halton import Halton
      6 from .korobov import Korobov

~/.conda-envs/playground/lib/python3.6/site-packages/qmcpy/discrete_distribution/sobol/__init__.py in <module>
----> 1 from .sobol import Sobol, DigitalNet

~/.conda-envs/playground/lib/python3.6/site-packages/qmcpy/discrete_distribution/sobol/sobol.py in <module>
      1 from .._discrete_distribution import DiscreteDistribution
      2 from ...util import ParameterError, ParameterWarning
----> 3 from ..c_lib import c_lib
      4 import ctypes
      5 from os.path import dirname, abspath, isfile

~/.conda-envs/playground/lib/python3.6/site-packages/qmcpy/discrete_distribution/c_lib/__init__.py in <module>
----> 1 from .load_c_lib import c_lib

~/.conda-envs/playground/lib/python3.6/site-packages/qmcpy/discrete_distribution/c_lib/load_c_lib.py in <module>
      3 from glob import glob
      4 
----> 5 c_lib = CDLL(glob(dirname(abspath(__file__))+'/c_lib*')[0], mode=RTLD_GLOBAL)

~/.conda-envs/playground/lib/python3.6/ctypes/__init__.py in __init__(self, name, mode, handle, use_errno, use_last_error)
    346 
    347         if handle is None:
--> 348             self._handle = _dlopen(self._name, mode)
    349         else:
    350             self._handle = handle

OSError: /home/gxia/.conda-envs/playground/lib/python3.6/site-packages/qmcpy/discrete_distribution/c_lib/c_lib.cpython-36m-x86_64-linux-gnu.so: undefined symbol: __svml_idiv4

**Comment by mikemccourt** [comment]
Hi Geng.  Sorry about the import issue.  Are you able to import this into python without the notebook, e.g., `python -c "from qmcpy.integrand import *"`?

**Reply by GengXia0406** [reply]
Hi Mike

I tried but the some error appears. Sea details below

;-----------------------------------------------------
(/home/gxia/.conda-envs/playground) [gxia@el1 ~]$ python -c "from qmcpy.integrand import *"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/gxia/.conda-envs/playground/lib/python3.6/site-packages/qmcpy/__init__.py", line 1, in <module>
    from .discrete_distribution import *
  File "/home/gxia/.conda-envs/playground/lib/python3.6/site-packages/qmcpy/discrete_distribution/__init__.py", line 4, in <module>
    from .sobol import Sobol, DigitalNet
  File "/home/gxia/.conda-envs/playground/lib/python3.6/site-packages/qmcpy/discrete_distribution/sobol/__init__.py", line 1, in <module>
    from .sobol import Sobol, DigitalNet
  File "/home/gxia/.conda-envs/playground/lib/python3.6/site-packages/qmcpy/discrete_distribution/sobol/sobol.py", line 3, in <module>
    from ..c_lib import c_lib
  File "/home/gxia/.conda-envs/playground/lib/python3.6/site-packages/qmcpy/discrete_distribution/c_lib/__init__.py", line 1, in <module>
    from .load_c_lib import c_lib
  File "/home/gxia/.conda-envs/playground/lib/python3.6/site-packages/qmcpy/discrete_distribution/c_lib/load_c_lib.py", line 5, in <module>
    c_lib = CDLL(glob(dirname(abspath(__file__))+'/c_lib*')[0], mode=RTLD_GLOBAL)
  File "/home/gxia/.conda-envs/playground/lib/python3.6/ctypes/__init__.py", line 348, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: /home/gxia/.conda-envs/playground/lib/python3.6/site-packages/qmcpy/discrete_distribution/c_lib/c_lib.cpython-36m-x86_64-linux-gnu.so: undefined symbol: __svml_idiv4
;-------------------------------------------------------------------

Thanks again for looking into this
Best regards
Geng Xia

**Reply by mikemccourt** [reply]
Thank you for your interest in QMCPy.  We'll see if there's something obvious that we can have you try.  Base case, I'm guessing that this is an issue with how we assume dynamic libraries were going to be linked.

While we're investigating, I'd like to ask a couple other questions.  Namely, 1) how did you find out about QMCPy and 2) what project are you working on where you would like to use it?  An important part of our development cycle is making sure we are prioritizing improvements that our users value.

**Reply by GengXia0406** [reply]
1) how did you find out about QMCPy 
I just google it. Recently, i want to use QMC for my research and i google to see it if there is a python package available.

2) what project are you working on where you would like to use it
I am an atmospheric science researcher and I want to use QMC for my parameter sensitivity experiments.

**Comment by jagadeesr** [comment]
Hi Geng,
Could you please add details of your OS and computer that you are running?

 << OSError:
/home/gxia/.conda-envs/playground/lib/python3.6/site-packages/qmcpy/discrete_distribution/c_lib/
c_lib.cpython-36m-x86_64-linux-gnu.so: undefined symbol: __svml_idiv4

The above error implies, "Short Vector Math Library" (libsvml.a) could not
be found.
Did you install qmcpy using pip or using the code from github?

On Fri, Apr 9, 2021 at 12:27 PM Michael McCourt ***@***.***>
wrote:

> Hi Gang. Sorry about the import issue. Are you able to import this into
> python without the notebook, e.g., python -c "from qmcpy.integrand import
> *"?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/QMCSoftware/QMCSoftware/discussions/244#discussioncomment-590919>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJTRVFZ4NOX5VNQAJ37FWLTH42INANCNFSM42UFY22Q>
> .
>


-- 
-Jags

**Reply by GengXia0406** [reply]
Hi Jagadeesr

I installed the program on a the NREL supercomputer. However, i am not an expert on this so not sure how to show you more details about the OS system. But i think that library should be installed.

I installed the program using pip, not from the source code. Should i try from the source code? I am very new to Github and would appreciate if you can provide more detail on how to pull and install using the source code. 

Below shows the details when i installed it using pip
;--------------------------------------------------------------------------------------------------
(/home/gxia/.conda-envs/playground) [gxia@el1 ~]$ pip install qmcpy
Requirement already satisfied: qmcpy in ./.conda-envs/playground/lib/python3.6/site-packages (1.0)
Requirement already satisfied: scipy>=1.2.0 in ./.conda-envs/playground/lib/python3.6/site-packages (from qmcpy) (1.5.3)
Requirement already satisfied: numpy>=1.18.5 in ./.conda-envs/playground/lib/python3.6/site-packages (from qmcpy) (1.19.5)
;--------------------------------------------------------------------------------------------------

Thanks again for looking into this
Best regards
Geng Xia

**Comment by jagadeesr** [comment]
Thank you, We were looking at easier solutions for you to get started.

You could get quickly started using this google Colab notebook while we
debug the issue.
Please make a copy of yourself.
https://colab.research.google.com/drive/1Nh6OhZ5w6trTFjyauIf0YAXD5JReyKpM?usp=sharing

To debug the error, please provide more details as below.
As per this log:
>> OSError:
/home/gxia/.conda-envs/playground/lib/python3.6/site-packages/qmcpy/discrete_distribution/c_lib/
c_lib.cpython-36m-x86_64-linux-gnu.so: undefined symbol: __svml_idiv4

It looks like the OS is Linux. Maybe the libsvml (or some other dynamic)
library is incompatible.
I see you installed qmcpy using: "pip install qmcpy"
Did you get any error or warning while doing it?

Could you please run these cmds to get your OS details for us to reproduce
the error?
>> cat /etc/os-release
>> lsb_release -a
>> hostnamectl
>> uname -a

Another alternative is to use docker, if you must run it on your local
computer.
Please see if you could run these cmds:
>> docker pull alegresor/qmcpy
>> docker run --rm -it --name=qmcpy  alegresor/qmcpy  /bin/bash


On Fri, Apr 9, 2021 at 4:57 PM GengXia-NREL ***@***.***>
wrote:

>
>    1.
>
>    how did you find out about QMCPy
>    I just google it. Recently, i want to use QMC for my research and i
>    google to see it if there is a python package available.
>    2.
>
>    what project are you working on where you would like to use it
>    I am an atmospheric science researcher and I want to use QMC for my
>    parameter sensitivity experiments.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/QMCSoftware/QMCSoftware/discussions/244#discussioncomment-591871>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJTRVAYK2N6EMTJMCKIT4DTH5Z4TANCNFSM42UFY22Q>
> .
>


-- 
-Jags

**Reply by GengXia0406** [reply]
Hi Jagadeesr 

Thank you. Below are the answers to your quetions 

1) I see you installed qmcpy using: "pip install qmcpy"
No, I did not receive any errors.  This shows the details when i installed it using pip
;--------------------------------------------------------------------------------------------------
(/home/gxia/.conda-envs/playground) [gxia@el1 ~]$ pip install qmcpy
Requirement already satisfied: qmcpy in ./.conda-envs/playground/lib/python3.6/site-packages (1.0)
Requirement already satisfied: scipy>=1.2.0 in ./.conda-envs/playground/lib/python3.6/site-packages (from qmcpy) (1.5.3)
Requirement already satisfied: numpy>=1.18.5 in ./.conda-envs/playground/lib/python3.6/site-packages (from qmcpy) (1.19.5)
;--------------------------------------------------------------------------------------------------

2a) cat /etc/os-release
;-----------------------------------------------------------------------
(/home/gxia/.conda-envs/playground) [gxia@el1 ~]$ cat /etc/os-release
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
;----------------------------------------------------------------

2.b) lsb_release -a
;-----------------------------------------------------
(/home/gxia/.conda-envs/playground) [gxia@el1 ~]$ lsb_release -a
LSB Version:	:core-4.1-amd64:core-4.1-noarch:cxx-4.1-amd64:cxx-4.1-noarch:desktop-4.1-amd64:desktop-4.1-noarch:languages-4.1-amd64:languages-4.1-noarch:printing-4.1-amd64:printing-4.1-noarch
Distributor ID:	CentOS
Description:	CentOS Linux release 7.7.1908 (Core)
Release:	7.7.1908
Codename:	Core
;-----------------------------------------------------

2.c) hostnamectl
;--------------------------------------------------------
(/home/gxia/.conda-envs/playground) [gxia@el1 ~]$ hostnamectl
   Static hostname: el1
         Icon name: computer-server
           Chassis: server
        Machine ID: 8c20756b80b249bbad9c1d60c90bbce9
           Boot ID: b7a6449529ef4d479c0d141b01b6095e
  Operating System: CentOS Linux 7 (Core)
       CPE OS Name: cpe:/o:centos:centos:7
            Kernel: Linux 3.10.0-1062.9.1.el7.x86_64
      Architecture: x86-64
;----------------------------------------------------------

2d) uname -a
;---------------------------------------------------------
(/home/gxia/.conda-envs/playground) [gxia@el1 ~]$ uname -a
Linux el1 3.10.0-1062.9.1.el7.x86_64 #1 SMP Fri Dec 6 15:49:49 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
;----------------------------------------------------------

3) I don't have 'docker' commend available on my local computer

Thanks 
Best regards
Geng Xia

**Comment by jagadeesr** [comment]
Hi GengXia,
Were you able to use this colab notebook?

> You could get quickly started using this google Colab notebook while we
> debug the issue.
> Please make a copy of yourself.
>
> https://colab.research.google.com/drive/1Nh6OhZ5w6trTFjyauIf0YAXD5JReyKpM?usp=sharing
>

It looks like your computer is using a different C compiler than what we
used so far.
QMCpy library has some code in C that gets compiled when you do pip install.
This seems to be producing very different results.
I will send you some more cmds to understand the type of compiler that you
are using.

On Mon, Apr 12, 2021 at 11:38 AM Jagadeeswaran Rathinavel <
***@***.***> wrote:

> Thank you, We were looking at easier solutions for you to get started.
>
> You could get quickly started using this google Colab notebook while we
> debug the issue.
> Please make a copy of yourself.
>
> https://colab.research.google.com/drive/1Nh6OhZ5w6trTFjyauIf0YAXD5JReyKpM?usp=sharing
>
> To debug the error, please provide more details as below.
> As per this log:
> >> OSError:
> /home/gxia/.conda-envs/playground/lib/python3.6/site-packages/qmcpy/discrete_distribution/c_lib/
> c_lib.cpython-36m-x86_64-linux-gnu.so: undefined symbol: __svml_idiv4
>
> It looks like the OS is Linux. Maybe the libsvml (or some other dynamic)
> library is incompatible.
> I see you installed qmcpy using: "pip install qmcpy"
> Did you get any error or warning while doing it?
>
> Could you please run these cmds to get your OS details for us to reproduce
> the error?
> >> cat /etc/os-release
> >> lsb_release -a
> >> hostnamectl
> >> uname -a
>
> Another alternative is to use docker, if you must run it on your local
> computer.
> Please see if you could run these cmds:
> >> docker pull alegresor/qmcpy
> >> docker run --rm -it --name=qmcpy  alegresor/qmcpy  /bin/bash
>
>
> On Fri, Apr 9, 2021 at 4:57 PM GengXia-NREL ***@***.***>
> wrote:
>
>>
>>    1.
>>
>>    how did you find out about QMCPy
>>    I just google it. Recently, i want to use QMC for my research and i
>>    google to see it if there is a python package available.
>>    2.
>>
>>    what project are you working on where you would like to use it
>>    I am an atmospheric science researcher and I want to use QMC for my
>>    parameter sensitivity experiments.
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/QMCSoftware/QMCSoftware/discussions/244#discussioncomment-591871>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAJTRVAYK2N6EMTJMCKIT4DTH5Z4TANCNFSM42UFY22Q>
>> .
>>
>
>
> --
> -Jags
>
>

-- 
-Jags

**Reply by GengXia0406** [reply]
Hi Jagadeesr

I am able to use the colab notebook. Thanks again for looking into this issue.
Best regards
Geng Xia

**Reply by mikemccourt** [reply]
If the collar is working for you, you can get started learning how QMCPy works there.  But, obviously, you'll need to be able to install it on your actual machine in order for it to run.

Can you tell me about your eventual desired setup?  Are you only trying to run QMCPy on the machine you are currently using?  Or are you developing this locally but then will eventually run this on a bigger machine at NREL?

**Reply by GengXia0406** [reply]
Hi Mike, the problem is resolved for me on NREL machine. However the solution is quiet interesting. 

After many tests, I realized that the qmcpy package only works (In my case) if the python version is 3.9. 
;--------------------------------------------
command1: conda create -n test python=3.9
command2: source activate test
command3: pip install qmcpy
;----------------------------------------------
 
Using similar commands, i will receive the exact same error with an earlier python version (i tested 3.8 and 3.7)

I don't know why this is the case, but i assume the python version will have an impact on the numpy and scipy that get installed. Or maybe there are some settings in the NREL machine that is causing this issue. I am not sure whether this is the similar case at your end. But i can use the package now. Thanks for all your helps 

Best regards
Geng Xia

**Reply by mikemccourt** [reply]
That is very interesting, and I do not know exactly what the problem is.  If I were to guess, I'm going to guess that the machine you are running on had python 3.7 and 3.8 already installed and, perhaps, the 3.9 installation was new.  Because it was new, there was no conflict between the libraries/compilers used to build numpy/scipy and those used to build qmcpy.  We will work to develop a consistent solution to this, but are glad that you are off and running.

Thank you for alerting us to this problem, and telling us about yourself and your research.  We would love to learn more as your project develops -- in particular, we always want to learn more about where QMC can be a beneficial tool.  As you write up your work and your project (both as an article and as publicly available code in GitHub), I hope that you would be willing to share this with us.

Thank you for using QMCPy.  Please reach out again if there is anything else that we can help you with.
-Mike, Jags, and the whole QMCPy team

**Comment by jagadeesr** [comment]
Great, you could start using the qmcpy while we debug.
Is google-colab good enough for your experiments?

If not and you still want to run this locally. Could you please
install docker using these instructions?
https://docs.docker.com/engine/install/centos/

Then please try:
>> docker pull alegresor/qmcpy
>> docker run --rm -it --name=qmcpy  alegresor/qmcpy  /bin/bash



On Tue, Apr 13, 2021 at 12:59 AM GengXia-NREL ***@***.***> wrote:
>
> Hi Jagadeesr
>
> I am able to use the colab notebook. Thanks again for looking into this issue.
> Best regards
> Geng Xia
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.



--
-Jags

**Reply by GengXia0406** [reply]
Hi Jagadeesr, the problem is resolved for me on NREL machine. However the solution is quiet interesting.

After many tests, I realized that the qmcpy package only works (In my case) if the python version is 3.9.
;--------------------------------------------
command1: conda create -n test python=3.9
command2: source activate test
command3: pip install qmcpy
;----------------------------------------------

Using similar commands, i will receive the exact same error with an earlier python version (i tested 3.8 and 3.7)

I don't know why this is the case, but i assume the python version will have an impact on the numpy and scipy that get installed. Or maybe there are some settings in the NREL machine that is causing this issue. I am not sure whether this is the similar case at your end. But i can use the package now. Thanks for all your helps

Best regards
Geng Xia
